File: apps/api/app/schemas/rfi.py
from datetime import date, datetime
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict

from app.schemas.common import PaginationMeta


class RfiCreate(BaseModel):
    subject: str
    question: str
    assigned_to: Optional[UUID] = None
    distribution_list: list[UUID] = []
    due_date: Optional[date] = None
    priority: str = "NORMAL"
    cost_impact: Optional[bool] = None
    schedule_impact: Optional[bool] = None
    drawing_reference: Optional[str] = None
    spec_section: Optional[str] = None
    location: Optional[str] = None


class RfiUpdate(BaseModel):
    subject: Optional[str] = None
    question: Optional[str] = None
    assigned_to: Optional[UUID] = None
    distribution_list: Optional[list[UUID]] = None
    due_date: Optional[date] = None
    priority: Optional[str] = None
    cost_impact: Optional[bool] = None
    schedule_impact: Optional[bool] = None
    drawing_reference: Optional[str] = None
    spec_section: Optional[str] = None
    location: Optional[str] = None


class RfiResponseCreate(BaseModel):
    """Official response to an RFI (not a comment)."""
    response: str
    attachments: list[UUID] = []


class RfiResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    project_id: UUID
    number: int
    formatted_number: str
    subject: str
    question: str
    official_response: Optional[str] = None
    status: str
    priority: str
    assigned_to: Optional[UUID] = None
    assigned_to_name: Optional[str] = None
    due_date: Optional[date] = None
    days_open: Optional[int] = None
    cost_impact: bool
    schedule_impact: bool
    drawing_reference: Optional[str] = None
    spec_section: Optional[str] = None
    location: Optional[str] = None
    created_by: Optional[UUID] = None
    created_by_name: Optional[str] = None
    responded_by: Optional[UUID] = None
    responded_by_name: Optional[str] = None
    responded_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime
    comments_count: int = 0


class RfiListResponse(BaseModel):
    data: list[RfiResponse]
    meta: PaginationMeta

File: apps/web/src/types/rfi.ts
export interface RFI {
  id: string;
  project_id: string;
  number: number;
  formatted_number: string;
  subject: string;
  question: string;
  official_response?: string | null;
  status: string;
  priority: string;
  assigned_to?: string | null;
  assigned_to_name?: string | null;
  due_date?: string | null;
  days_open?: number | null;
  cost_impact: boolean;
  schedule_impact: boolean;
  drawing_reference?: string | null;
  spec_section?: string | null;
  location?: string | null;
  created_by?: string | null;
  created_by_name?: string | null;
  responded_by?: string | null;
  responded_by_name?: string | null;
  responded_at?: string | null;
  created_at: string;
  updated_at: string;
  comments_count: number;
}

export interface RFICreateInput {
  subject: string;
  question: string;
  assigned_to?: string;
  due_date?: string;
  priority?: string;
  cost_impact?: boolean;
  schedule_impact?: boolean;
}

export interface RFIUpdateInput {
  subject?: string;
  question?: string;
  assigned_to?: string;
  due_date?: string;
  priority?: string;
  cost_impact?: boolean;
  schedule_impact?: boolean;
}
